traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/ds/LetsEncryptDnsChallengeWatcher.java

```diff
@@ -0,0 +1,180 @@
+    private static final String configFile = "/opt/traffic_router/db/cr-config.json";
+
+    private ConfigHandler configHandler;
+
+    public void setConfigHandler(final ConfigHandler configHandler) {
+        this.configHandler = configHandler;
+    }
+
+    public LetsEncryptDnsChallengeWatcher() {
+        setDatabaseUrl(DEFAULT_LE_DNS_CHALLENGE_URL);
+    }
+
+    @Override
+    public boolean useData(final String data) {
+        try {
+            final ObjectMapper mapper = new ObjectMapper(new JsonFactory());
+            final HashMap<String, List<LetsEncryptDnsChallenge>> dataMap = mapper.readValue(data, new TypeReference<HashMap<String, List<LetsEncryptDnsChallenge>>>() { });
+            final List<LetsEncryptDnsChallenge> challengeList = dataMap.get("response");
+
+            final JsonNode mostRecentConfig = mapper.readTree(readConfigFile());
+            final ObjectNode deliveryServicesNode = (ObjectNode) JsonUtils.getJsonNode(mostRecentConfig, ConfigHandler.deliveryServicesKey);
+
+
+            challengeList.forEach(challenge -> {
+                final StringBuilder sb = new StringBuilder();
+                sb.append(challenge.getFqdn());
+                if (!challenge.getFqdn().endsWith(".")) {
+                    sb.append('.');
+                }
+                final String challengeDomain = sb.toString();
+                final String fqdn = challengeDomain.substring(0, challengeDomain.length() - 1).replace("_acme-challenge.", "");
+
+                ObjectNode deliveryServiceConfig = null;
+                String dsLabel = "";
+                final StringBuilder nameSb = new StringBuilder();
+                nameSb.append("_acme-challenge");
+                for (final String label : fqdn.split("\\.")) {
+                    deliveryServiceConfig = (ObjectNode) deliveryServicesNode.get(label);
+                    if (deliveryServiceConfig != null) {
+                        dsLabel = label;
+                        break;
+                    } else {
+                        nameSb.append('.');
+                        nameSb.append(label);
+                    }
+                }
+
+                final String name = nameSb.toString();
+
+                final ArrayNode staticDnsEntriesNode = updateStaticEntries(challenge, name, mapper, deliveryServiceConfig);
+
+                deliveryServiceConfig.set("staticDnsEntries", staticDnsEntriesNode);
+                deliveryServicesNode.set(dsLabel, deliveryServiceConfig);
+
+            });
+
+            final ObjectNode statsNode = (ObjectNode) mostRecentConfig.get("stats");
+            statsNode.put("date", Instant.now().toEpochMilli() / 1000L);
+
+            final ObjectNode fullConfig = (ObjectNode) mostRecentConfig;
+            fullConfig.set(ConfigHandler.deliveryServicesKey, deliveryServicesNode);
+            fullConfig.set("stats", statsNode);
+
+            try {
+                configHandler.processConfig(fullConfig.toString());
```

Comment:
All these Traffic Router changes look good to me. This call to 'processConfig' currently has a high impact on the Traffic Router service because it forces TR to reload its entire configuration and re-sign all the DNS zones on the CDN, but the [Delivery Service Only Snapshots 2785](https://github.com/apache/trafficcontrol/pull/2785) feature which will soon be added to Traffic Router will almost completely mitigate this impact. I recommend the these two PRs be placed in the same release. 
I recommend